solutions/USACO feb16-milk-pails.mdx

```diff
@@ -2,15 +2,99 @@
 id: usaco-620
 source: USACO Silver 2016 February
 title: Milk Pails
-author: Óscar Garries
+author: Óscar Garries, Neo Wang
 ---
 
 [Official Analysis](http://www.usaco.org/current/data/sol_pails_silver_feb16.html)
 
 <LanguageSection>
 <CPPSection>
 
-## C++ Implementation
+## C++ Implementation (BFS)
+
+We can directly simulate all operations using BFS.
+
+When we are performing our BFS, we only have to keep track of the amount of operations used so far for the current amount of milk in pails $X$ and $Y$. That way, we can check all nodes with distance $\le K$ once our BFS is finished, and then compute the smallest value of  $|i + j -M|$.
+
+In our matrix, the current amount of milk in $X$ is $i$, and $Y$ is $j$.
+
+For every step, we simulate all the possible actions.
+
+1. Fill either pail completely to the top: set $i$ to $X$ or $j$ to $Y$.
+2. Empty either pail: set $i$ or $j$ to $0$.
+3. Pour the contents of one pail into another without overflowing or running out of milk.
+
+To perform operation 3, let the quantity to be poured $Q$ be denoted by either $Q_x=\min(i, X - j)$ or $Q_y=\min(j, Y-i)$, depending on which bucket is poured ($Q_x$ if we are pouring out of $x$ and vice versa). Then, to pour $X$ into $Y$, we subtract this poured quantity $Q_x$ from $X$ and add it to $Y$. In other words, if we were to pour $X$ into $Y$, the new amount in $X$ would be $X - Q_x$, and the new amount in $Y$ is $Y + Q_x$. To pour $Y$ into $X$, this would be $Y_{\textrm{new}}=Y-Q_y$, $X_{\textrm{new}}=X+Q_y$.
```

Comment:
```suggestion
To perform operation 3, let the quantity to be poured $Q$ be denoted by either $Q_x=\min(i, Y - j)$ or $Q_y=\min(j, X-i)$, depending on which bucket is poured ($Q_x$ if we are pouring out of $x$ and vice versa). Then, to pour $X$ into $Y$, we subtract this poured quantity $Q_x$ from $X$ and add it to $Y$. In other words, if we were to pour $X$ into $Y$, the new amount in $X$ would be $X - Q_x$, and the new amount in $Y$ is $Y + Q_x$. To pour $Y$ into $X$, this would be $Y_{\textrm{new}}=Y-Q_y$, $X_{\textrm{new}}=X+Q_y$.
```

Reply:
typo?

Reply:
oops

Reply:
Nice catch!